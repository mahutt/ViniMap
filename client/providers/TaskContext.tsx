import { Task } from '@/types';
import { storage } from '@/services/StorageService';
import { createContext, useState, useContext, ReactNode, useEffect, useMemo } from 'react';

type TaskContextType = {
  selectedTasks: Task[];
  setSelectedTasks: (tasks: Task[]) => void;
  markTask: (taskId: string, completionState: boolean) => void;
  tasks: Task[];
  setTasks: React.Dispatch<React.SetStateAction<Task[]>>;
};

const TaskContext = createContext<TaskContextType | undefined>(undefined);

export const TaskProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [tasks, setTasks] = useState<Task[]>(() => {
    try {
      const savedTasksString = storage.getString('allTasks');

      if (!savedTasksString) {
        return [];
      }

      const savedTasks: Task[] = JSON.parse(savedTasksString);
      savedTasks.forEach((task) => {
        if (task.startTime) {
          task.startTime = new Date(task.startTime);
        }
      });
      return savedTasks;
    } catch (error) {
      console.error('Error loading tasks from storage:', error);
      return [];
    }
  });

  const [selectedTasks, setSelectedTasks] = useState<Task[]>([]);

  useEffect(() => {
    try {
      storage.set('allTasks', JSON.stringify(tasks));
    } catch (error) {
      console.error('Error saving tasks:', error);
    }
  }, [tasks]);

  const markTask = (taskId: string, completionState: boolean) => {
    setSelectedTasks((prevTasks) =>
      prevTasks.map((task) => (task.id === taskId ? { ...task, completed: completionState } : task))
    );
  };

  const contextValue = useMemo(
    () => ({
      selectedTasks,
      setSelectedTasks,
      markTask,
      tasks,
      setTasks,
    }),
    [selectedTasks, tasks]
  );

  return <TaskContext.Provider value={contextValue}>{children}</TaskContext.Provider>;
};

export const useTask = () => {
  const context = useContext(TaskContext);
  if (!context) {
    throw new Error('useTask must be used within a TaskProvider');
  }
  return context;
};
