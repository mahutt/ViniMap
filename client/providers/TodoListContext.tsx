import { Task } from '@/modules/map/Types';
import { createContext, useState, useContext, ReactNode } from 'react';

type TaskContextType = {
  selectedTasks: Task[];
  setSelectedTasks: (tasks: Task[]) => void;
};

const TaskContext = createContext<TaskContextType | undefined>(undefined);

export const TaskProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [selectedTasks, setSelectedTasks] = useState<Task[]>([]);

  return (
    <TaskContext.Provider value={{ selectedTasks, setSelectedTasks }}>
      {children}
    </TaskContext.Provider>
  );
};

export const useTask = () => {
  const context = useContext(TaskContext);
  if (!context) {
    throw new Error('useTask must be used within a TaskProvider');
  }
  return context;
};
