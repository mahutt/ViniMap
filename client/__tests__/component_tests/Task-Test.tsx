import React from 'react';
import { render } from '@testing-library/react-native';
import { TaskProvider } from '@/providers/TodoListContext'; // Import TaskProvider
import TasksScreen from '@/app/(tabs)/tasks';

test('renders TasksScreen correctly', () => {
  const { toJSON } = render(
    <TaskProvider>
      {' '}
      {}
      <TasksScreen />
    </TaskProvider>
  );

  expect(toJSON()).toMatchSnapshot();
});

test('displays "No tasks yet" when task list is empty', () => {
  const { getByText } = render(
    <TaskProvider>
      <TasksScreen />
    </TaskProvider>
  );

  expect(getByText('No tasks yet.')).toBeTruthy();
});
test('renders the add task button', () => {
  const { getByText } = render(
    <TaskProvider>
      <TasksScreen />
    </TaskProvider>
  );

  expect(getByText('+')).toBeTruthy(); // Checking for the "+" button
});

test('renders the generate path button', () => {
  const { getByText } = render(
    <TaskProvider>
      <TasksScreen />
    </TaskProvider>
  );

  expect(getByText('Generate Path')).toBeTruthy();
});
