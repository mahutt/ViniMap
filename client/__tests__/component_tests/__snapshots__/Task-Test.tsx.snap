// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TasksScreen correctly 1`] = `
[
  " ",
  <View
    style={
      {
        "backgroundColor": "#E8EAED",
        "flex": 1,
        "padding": 20,
      }
    }
  >
    <View
      style={
        {
          "paddingTop": 80,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingBottom": 20,
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 24,
              "fontWeight": "bold",
            }
          }
        >
          Tasks
        </Text>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <ViewManagerAdapter_SymbolModule
            animated={false}
            colors={[]}
            name="arrow.uturn.backward"
            resizeMode="scaleAspectFit"
            style={
              [
                {
                  "height": 24,
                  "width": 24,
                },
                [
                  {
                    "height": 24,
                    "width": 24,
                  },
                  undefined,
                ],
              ]
            }
            tintColor="#852C3A"
            type="monochrome"
            weight="regular"
          />
        </View>
      </View>
      <RCTScrollView
        style={
          {
            "backgroundColor": "#E8EAED",
            "borderColor": "#852C3A",
            "borderRadius": 10,
            "borderWidth": 2,
            "height": 500,
            "paddingHorizontal": 10,
            "paddingVertical": 10,
          }
        }
      >
        <View>
          <Text
            style={
              {
                "color": "gray",
                "marginTop": 20,
                "textAlign": "center",
              }
            }
          >
            No tasks yet.
          </Text>
        </View>
      </RCTScrollView>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "column",
            "justifyContent": "center",
            "marginTop": 10,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#852C3A",
              "borderRadius": 10,
              "height": 50,
              "justifyContent": "center",
              "marginBottom": 5,
              "opacity": 1,
              "width": 50,
            }
          }
        >
          <Text
            style={
              {
                "color": "#FFF",
                "fontSize": 24,
              }
            }
          >
            +
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#852C3A",
              "borderRadius": 10,
              "height": 45,
              "justifyContent": "center",
              "opacity": 1,
              "width": 250,
            }
          }
        >
          <Text
            style={
              {
                "color": "#FFF",
                "fontSize": 18,
              }
            }
          >
            Generate Path
          </Text>
        </View>
      </View>
    </View>
  </View>,
]
`;
